File: src/main/java/net/vanillaoutsider/betterdogs/ai/WolfStormAnxietyGoal.java
// Copyright (C) 2026 Dasik (Rifaditya) | GNU GPLv3
// Verified against: Minecraft 26.2
package net.vanillaoutsider.betterdogs.ai;

import net.minecraft.core.BlockPos;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.ai.goal.Goal;
import net.minecraft.world.entity.ai.util.DefaultRandomPos;
import net.minecraft.world.entity.animal.wolf.Wolf;
import net.minecraft.world.phys.Vec3;
import net.vanillaoutsider.betterdogs.mixin.WolfAccessor;
import net.vanillaoutsider.betterdogs.util.WolfStormHelper;

import java.util.EnumSet;

/**
 * Dedicated single-purpose AI goal for thunderstorm anxiety, shelter pathing, and whimpering reactions.
 */
public class WolfStormAnxietyGoal extends Goal {

    private final Wolf wolf;

    public WolfStormAnxietyGoal(Wolf wolf) {
        this.wolf = wolf;
        this.setFlags(EnumSet.of(Goal.Flag.MOVE, Goal.Flag.LOOK));
    }

    @Override
    public boolean canUse() {
        if (!WolfStormHelper.isStormAnxietyActive(this.wolf)) {
            return false;
        }
        float multiplier = WolfStormHelper.getPersonalityMultiplier(this.wolf);
        if (multiplier <= 0.0f) {
            return false;
        }
        return this.wolf.getRandom().nextFloat() < (0.25f * multiplier);
    }

    @Override
    public boolean canContinueToUse() {
        return WolfStormHelper.isStormAnxietyActive(this.wolf);
    }

    @Override
    public void start() {
        if (!this.wolf.isOrderedToSit()) {
            BlockPos shelterPos = WolfStormHelper.findShelterTarget(this.wolf);
            if (shelterPos != null) {
                this.wolf.getNavigation().moveTo(shelterPos.getX() + 0.5, shelterPos.getY(), shelterPos.getZ() + 0.5, 1.25);
            } else {
                Vec3 target = DefaultRandomPos.getPos(this.wolf, 8, 4);
                if (target != null) {
                    this.wolf.getNavigation().moveTo(target.x, target.y, target.z, 1.0);
                }
            }
        }
    }

    @Override
    public void tick() {
        if (this.wolf.getRandom().nextFloat() < 0.05F) {
            try {
                var soundSet = ((WolfAccessor) this.wolf).betterdogs$invokeGetSoundSet();
                if (soundSet != null && soundSet.whineSound() != null) {
                    this.wolf.playSound(soundSet.whineSound().value(), 1.0F, 1.0F);
                }
            } catch (Exception ignored) {
            }
        }

        // Particle feedback while outdoors in rain/thunder
        if (this.wolf.level() instanceof ServerLevel serverLevel) {
            if (this.wolf.level().canSeeSky(this.wolf.blockPosition()) && this.wolf.getRandom().nextFloat() < 0.15F) {
                serverLevel.sendParticles(ParticleTypes.SPLASH, this.wolf.getRandomX(0.8), this.wolf.getRandomY() + 0.3, this.wolf.getRandomZ(0.8), 2, 0.1, 0.1, 0.1, 0.02);
            }
        }

        if (!this.wolf.isOrderedToSit() && this.wolf.getNavigation().isDone()) {
            BlockPos shelterPos = WolfStormHelper.findShelterTarget(this.wolf);
            if (shelterPos != null) {
                this.wolf.getNavigation().moveTo(shelterPos.getX() + 0.5, shelterPos.getY(), shelterPos.getZ() + 0.5, 1.25);
            } else {
                Vec3 target = DefaultRandomPos.getPos(this.wolf, 8, 4);
                if (target != null) {
                    this.wolf.getNavigation().moveTo(target.x, target.y, target.z, 1.0);
                }
            }
        }

        if (this.wolf.getRandom().nextFloat() < 0.2F) {
            double spread = 4.0;
            this.wolf.getLookControl().setLookAt(
                    this.wolf.getX() + (this.wolf.getRandom().nextDouble() - 0.5) * spread,
                    this.wolf.getEyeY(),
                    this.wolf.getZ() + (this.wolf.getRandom().nextDouble() - 0.5) * spread,
                    10.0F,
                    (float) this.wolf.getMaxHeadXRot()
            );
        }
    }

    @Override
    public void stop() {
        this.wolf.getNavigation().stop();
    }
}

File: src/main/kotlin/net/vanillaoutsider/betterdogs/ai/WolfStormAnxietyGoal.kt
package net.vanillaoutsider.betterdogs.ai

import net.minecraft.world.entity.ai.goal.Goal
import net.minecraft.world.entity.animal.wolf.Wolf
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundEvents
import net.minecraft.resources.Identifier
import net.minecraft.world.entity.ai.util.DefaultRandomPos
import net.vanillaoutsider.betterdogs.config.BetterDogsConfig
import java.util.EnumSet

/**
 * AI Goal for Wolf anxiety during thunderstorms.
 * - Whines occasionally.
 * - Shakes (uses wet dog shake animation as fear response).
 * - Paces around nervously (if not sitting).
 */
class WolfStormAnxietyGoal(private val wolf: Wolf) : Goal() {

    init {
        this.flags = EnumSet.of(Flag.MOVE, Flag.LOOK)
    }

    override fun canUse(): Boolean {
        // Only trigger if tamed
        if (!wolf.isTame) return false

        // Only during thunderstorms
        if (!wolf.level().isThundering) return false

        // Only if exposed to sky? or checking for skylight?
        // For simpler "fear of thunder", just the global weather is enough.
        // But maybe simple check to see if we aren't deep underground (optional, but keep simple for now)
        
        // Random chance to start being anxious (1% per tick)
        return wolf.random.nextFloat() < BetterDogsConfig.get().stormAnxietyTriggerChance
    }

    override fun canContinueToUse(): Boolean {
        // Continue as long as it thundering and we haven't finished our "anxiety episode"
        return wolf.level().isThundering && wolf.random.nextFloat() < 0.98 // Small chance to stop
    }

    override fun start() {
        // If not sitting, maybe pick a random spot to pace to
        if (!wolf.isOrderedToSit) {
            val target = DefaultRandomPos.getPos(wolf, 3, 2)
            if (target != null) {
                wolf.navigation.moveTo(target.x, target.y, target.z, 1.0)
            }
        }
    }

    override fun tick() {
        val config = BetterDogsConfig.get()
        // 1. Whining Sound (Low chance)
        if (wolf.random.nextFloat() < config.stormWhineChance) { // 2% chance per tick
            // Try standard GENERIC_HURT if available
            wolf.playSound(net.minecraft.sounds.SoundEvents.GENERIC_HURT, 1.0f, 2.0f)
        }

        // 2. Shaking (Visual fear)
        // We use the "wet" flag or similar to trigger shake if possible, or just subtle look changes.
        // Actually, invoking the full "shake off water" animation might be too much (interrupts movement).
        // Let's stick to sound + movement for now, maybe look around randomly.
        
        if (wolf.random.nextFloat() < 0.05) {
             wolf.lookControl.setLookAt(
                 wolf.x + (wolf.random.nextDouble() - 0.5) * 10, 
                 wolf.eyeY, 
                 wolf.z + (wolf.random.nextDouble() - 0.5) * 10, 
                 10.0f, 
                 wolf.maxHeadXRot.toFloat()
             )
        }
    }

    override fun stop() {
        wolf.navigation.stop()
    }
}
